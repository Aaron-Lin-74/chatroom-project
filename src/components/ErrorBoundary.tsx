import { Component, ErrorInfo, ReactNode } from 'react';

interface Props {
  children: ReactNode;
}

interface State {
  hasError: boolean;
}
export default class ErrorBoundary extends Component<Props, State> {
  constructor(props: Props) {
    super(props);
    this.state = {
      hasError: false,
    };
  }

  public static getDerivedStateFromError(_: Error): State {
    return {
      hasError: true,
    };
  }

  public componentDidCatch(error: Error, errorInfo: ErrorInfo) {
    console.log('Uncaught error:', error, errorInfo);
  }

  public render() {
    if (this.state.hasError) {
      return (
        <div className='app-error-container'>
          <h1>Oops! Something went wrong.</h1>
          <h2>Please try later again.</h2>
        </div>
      );
    }
    return this.props.children;
  }
}
