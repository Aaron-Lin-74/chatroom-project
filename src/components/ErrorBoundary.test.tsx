import { screen, render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom';
import ErrorBoundary from './ErrorBoundary';

describe('Test suites for ErrorBoundary component', () => {
  const errorObject = console.error;
  const logObject = console.log;
  beforeEach(() => {
    console.error = jest.fn();
    console.log = jest.fn();
  });

  afterEach(() => {
    console.error = errorObject;
    console.log = logObject;
  });
  const mockChildren = <h1>Mock children</h1>;
  function Bomb({ shouldThrow }: { shouldThrow: boolean }) {
    if (shouldThrow) {
      throw new Error('An error I made');
    } else {
      return null;
    }
  }
  test('should match the snapshot', () => {
    const { asFragment } = render(
      <ErrorBoundary>
        <Bomb shouldThrow={true} />
      </ErrorBoundary>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test('should render the children component', () => {
    render(<ErrorBoundary children={mockChildren} />);
    screen.getByRole('heading', { name: 'Mock children' });
  });

  test('should render the error page when a component error occurs', () => {
    render(
      <ErrorBoundary>
        <Bomb shouldThrow={true} />
      </ErrorBoundary>
    );
    screen.getByRole('heading', { name: 'Oops! Something went wrong.' });
    screen.getByRole('heading', { name: 'Please try later again.' });
  });

  test('should be able to try again, and remove the error content', () => {
    const { rerender } = render(
      <ErrorBoundary>
        <Bomb shouldThrow={true} />
        {mockChildren}
      </ErrorBoundary>
    );
    // state is prever
    rerender(<ErrorBoundary children={mockChildren} />);
    userEvent.click(screen.getByText(/try again/i));
    expect(
      screen.queryByRole('heading', { name: 'Oops! Something went wrong.' })
    ).not.toBeInTheDocument();
    expect(screen.queryByText(/try again/i)).not.toBeInTheDocument();
  });
});
