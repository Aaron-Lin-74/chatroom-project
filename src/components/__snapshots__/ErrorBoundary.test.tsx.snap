// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test suites for ErrorBoundary component should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="app-error-container"
  >
    <h1>
      Oops! Something went wrong.
    </h1>
    <h2>
      Please try later again.
    </h2>
  </div>
</DocumentFragment>
`;
